core/node/bitswap.go

```diff
@@ -40,13 +45,34 @@ func OnlineExchange(cfg *config.Config, provide bool) interface{} {
 			bitswap.EngineTaskWorkerCount(int(internalBsCfg.EngineTaskWorkerCount.WithDefault(DefaultEngineTaskWorkerCount))),
 			bitswap.MaxOutstandingBytesPerPeer(int(internalBsCfg.MaxOutstandingBytesPerPeer.WithDefault(DefaultMaxOutstandingBytesPerPeer))),
 		}
-		exch := bitswap.New(helpers.LifecycleCtx(mctx, lc), bitswapNetwork, bs, opts...)
+
+		return bitswapOptionsOut{BitswapOpts: opts}
+	})
+}
+
+type onlineExchangeIn struct {
+	fx.In
+
+	Mctx        helpers.MetricsCtx
+	Host        host.Host
+	Rt          irouting.TieredRouter
+	Bs          blockstore.GCBlockstore
+	BitswapOpts []bitswap.Option `group:"bitswap-options"`
+}
+
+// OnlineExchange creates new LibP2P backed block exchange (BitSwap).
+// Additional options to bitswap.New can be provided via the "bitswap-options"
+// group.
+func OnlineExchange() fx.Option {
+	return fx.Provide(func(in onlineExchangeIn, lc fx.Lifecycle) exchange.Interface {
```

Comment:
I think I saw somewhere that `fx.Lifecycle` was provided outside of an `In` struct. Let me know if that's the right way or not...

Reply:
why are you moving `fx.Provide` inside the function?

Reply:
Because I asked for it, I I like type safety (returning `fx.Option` is imo clearer than `interface{}`).

Maybe it's less clear because you don't really what each option do (then we should rename it to `ProvideOnlineExchange` maybe ?)

Reply:
I do not disagree with you, but everywhere else `fx.Provide` is called outside.

Reply:
I changed this back to have `fx.Provide` outside the function.